File: DS/Linked LIst/DoublyLinkedList.java
class Node {
    int data;
    Node next;
    Node prev;
}

class Runner{
    
    public static void main(String[] args) {
        LinkedList list = new LinkedList();
        list.insert(5);
        list.insert(10);
        list.insert(20);
        list.insertAt(0,1);
        list.insertAt(3,15);
      //  list.insertAt(5,25);
       // list.deleteAt(3);
        //list.reverseList();
        list.show();
        System.out.println();
        list.search(35);
        //list.getHead();
        //list.getTail();
    }
    
}

class LinkedList {
    
    Node head, tail;
    public void insert(int data) {
        Node newnode = new Node();
        newnode.data = data;
        newnode.prev = null;
        newnode.next = null;
        
        if(head == null) {
            head = tail = newnode;
        }
        else {
            tail.next = newnode;
            newnode.prev = tail;
            tail = newnode;
        }
    }
    
    public void insertAtBegin(int data) {
        Node newnode = new Node();
        newnode.data = data;
        newnode.prev = null;
        newnode.next = head;
        head.prev = newnode;
        head = newnode;
    }
    
    public void insertAt(int index,int data) {
        if(index < 0 || index > size()){
            System.out.println("Invalid index..");
            return;
        }
        if(index == 0){
            insertAtBegin(data);
        }
        else {
            Node newnode = new Node();
            newnode.data = data;
            Node tempnode = head;
            
           
            for(int i = 0; i < index - 1; i++) {
                tempnode = tempnode.next;
            }
            
             if(index == size()){
                tempnode.next = newnode;
                newnode.prev = tempnode;
                tail = newnode;
                return;
            }
            newnode.next = tempnode.next;
            (tempnode.next).prev = newnode;
            newnode.prev = tempnode;
            tempnode.next = newnode;
        }
    }
    
    public void deleteAt(int index) {
        if(index < 0 || index >= size()){
            System.out.println("Invalid index..");
            return;
        }
        if(index == 0)
            head = head.next;
        else { 
            Node tempnode = head;
            for(int i = 0; i < index - 1 ; i++) {
                tempnode = tempnode.next;
            }
            
            if(index == size()-1){
                tail = tail.prev;
                tail.next = null;
                return;
            }
            (tempnode.next).prev = tempnode;
            tempnode.next = (tempnode.next).next;
        }
    }
    
    public void reverseList() {
        
        //Using tail pointer.
      /*Node tempnode = tail;
        
        if(head == null) {
            System.out.println("List is empty..");
            return;
        }
        
        System.out.print("Reversed List : ");
        while(tempnode != null) {
            System.out.print(tempnode.data + " ");
            tempnode = tempnode.prev;
        }*/
        
        // Without tail pointer
        Node currnode = head, nextnode = null;
        while(currnode != null) {
            nextnode = currnode.next;
            currnode.next = currnode.prev;
            currnode.prev = nextnode;
            currnode = nextnode;
        }
        currnode = head;
        head = tail;
        tail = currnode;
        
    }
    
    public void search(int data) {
	    if(head == null) {
	        System.out.println("List is empty.");
	        return;
	    }
	    Node tempnode = head;
	    while(tempnode != null){
	        if(tempnode.data == data){
	            System.out.println("Data found..");
	            return;
	        }
	        tempnode = tempnode.next;
	    }
	    System.out.println("Data not found..");
	}
    
    public void show() {
        if(head == null) {
            System.out.println("List is empty..");
            return;
        }
        Node tempnode = head;
        System.out.println();
        while(tempnode != null) {
            System.out.print(tempnode.data + " ");
            tempnode = tempnode.next;
        }
    }
    
    public int size() {
        int count = 0;
        Node tempnode = head;
        while(tempnode != null) {
            count++;
            tempnode = tempnode.next;
        }
        return count;
    }
    
    public void getHead() {
        System.out.println("\nHead is : " + head.data);
    }
    
    public void getTail() {
        System.out.println("Tail is : " + tail.data);
    }
}

File: DS/Python/Linked List/DoublyLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None

    def insertAtBegining(self, data):
        new_node = Node(data)
        new_node.next = self.head
        if self.head is not None:
            self.head.prev = new_node
        self.head = new_node

    def insertAtEnd(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        else:
            temp = self.head
            while temp.next:
                temp = temp.next
            temp.next = new_node
            new_node.prev = temp

    def insertAt(self, pos, data):
        new_node = Node(data)
        if pos < 0 or pos > self.size():
            raise IndexError("Invalid index!")
        elif pos == 0:
            self.insertAtBegining(data)
        elif pos == self.size():
            self.insertAtEnd(data)
        else:
            i = 0
            temp = self.head
            while i < pos - 1:
                temp = temp.next
                i += 1
            new_node.next = temp.next
            temp.next.prev = new_node
            new_node.prev = temp
            temp.next = new_node

    def remove(self, data):
        if self.head is None:
            raise ValueError("List is empty!")
        elif self.head.data == data:
            self.head = self.head.next
        else:
            flag = False
            temp = self.head
            while temp:
                if temp.data == data:
                    flag = True
                    break
                prev = temp
                temp = temp.next
            if flag:
                prev.next = temp.next
                if temp.next is not None:
                    temp.next.prev = prev
            else:
                raise ValueError("Item not in list!")

    def printList(self):
        if self.head is None:
            print("List is empty!")
        else:
            temp = self.head
            while temp:
                print(temp.data, end=" ")
                temp = temp.next

    def size(self):
        count = 0
        if self.head:
            temp = self.head
            while temp:
                count += 1
                temp = temp.next
        return count


dl = DoublyLinkedList()
dl.insertAt(0, 10)
dl.insertAt(1, 20)
dl.insertAtEnd(30)
dl.insertAtEnd(40)
dl.insertAtBegining(5)
dl.remove(5)

dl.printList()
